File: src/main/python/basic/queue.py
# -*- coding: utf-8 -*-

class Queue:
    def __init__(self):
        self.items = []
        self._sizes = 0

    def enqueue(self, item):
        # None can be added to queue
        self.items.insert(0, item)
        self._sizes += 1

    def dequeue(self):
        if self.is_empty():
            raise Exception("queue is empty")
        item = self.items.pop()
        self._sizes -= 1
        return item

    def size(self):
        return self._sizes

    def is_empty(self):
        return self._sizes == 0

def test_queue():
    queue = Queue()
    queue.enqueue(1)
    queue.enqueue(4)
    queue.enqueue("adolph")

    assert queue.size() == 3
    assert queue.dequeue() == 1
    assert queue.size() == 2

File: languages/golang/basic/queue.go
package basic

import "log"

type Queue struct {
	items	[]interface{}
	size	uint64
}

func NewQueue() *Queue {
	q := &Queue{
		items: make([]interface{}, 0),
		size: 0,
	}
	return q
}

func (q *Queue) Enqueue(item interface{}) {
	if item == nil {
		log.Fatal("argument for Enqueue() should not nil")
	}
	q.items = append(q.items, item)
	q.size++
}

func (q *Queue) Dequeue() interface{} {
	if q.IsEmpty() {
		log.Fatal("Queue is empty")
	}
	item := q.items[0]
	q.items = q.items[1:]
	q.size--
	return item
}

func (q *Queue) IsEmpty() bool {
	return q.size == 0
}

func (q *Queue) Size() uint64 {
	return q.size
}
